tiledb/sm/misc/uuid.cc

```diff
@@ -28,103 +28,83 @@
  * @section DESCRIPTION
  *
  * This file defines a platform-independent UUID generator.
+ *
+ * Initially from: https://gist.github.com/kvelakur/9069c9896577c3040030
+ * "Generating a Version 4 UUID using OpenSSL"
  */
 
+#include <cstdio>
 #include <mutex>
 #include <vector>
 
 #include "tiledb/sm/misc/uuid.h"
 
 #ifdef _WIN32
-#include <Rpc.h>
+#include <windows.h>
+#ifndef NT_SUCCESS
+#define NT_SUCCESS(Status) (((NTSTATUS)(Status)) >= 0)
+#endif
+#ifndef BCRYPT_RNG_ALG_HANDLE
+#define BCRYPT_RNG_ALG_HANDLE ((BCRYPT_ALG_HANDLE)0x00000081)
+#endif
 #else
 #include <openssl/err.h>
 #include <openssl/rand.h>
-#include <cstdio>
 #endif
 
-using namespace tiledb::common;
+namespace tiledb::sm::uuid {
 
-namespace tiledb {
-namespace sm {
-namespace uuid {
+struct Uuid {
```

Comment:
This whole UUID struct seems fairly overboard for generating a UUID. I'd simplify it to just two functions for generating bytes and formatting them as hex. Since we've put this in the `uuid` namespace, calling code would just need to use `auto uuid = uuid::generate();` or similar. Might be you'll want to move the rand_bytes function to some other place but I don't think we have any custom rand modules yet so unless you know of one I wouldn't bother creating it.

```suggestion
#ifdef _WIN32

void rand_bytes(std::string& bytes, size_t num_bytes) {
  bytes.resize(num_bytes);
  int rc = BCryptGenRandom(
      BCRYPT_RNG_ALG_HANDLE,
      reinterpret_cast<uint8_t*>(bytes.data()),
      num_bytes,
      0);
  if (!NT_SUCCESS(rc)) {
    throw std::runtime_error("Cannot generate GUID: BCryptGenRandom failed.");
  }
}

#else

void rand_bytes(std::string& bytes, size_t num_bytes) {
  bytes.resize(num_bytes);
  int rc = RAND_bytes(reinterpret_cast<uint8_t*>(bytes.data()), num_bytes);
  if (rc < 1) {
    char err_msg[256];
    ERR_error_string_n(ERR_get_error(), err_msg, sizeof(err_msg));
    throw std::runtime_error(
        "Cannot generate random bytes with OpenSSL: " + std::string(err_msg));
  }
}

#endif

std::string generate() {
  std::string bytes(16);
  rand_bytes(bytes, 16);

  std::string uuid(32);
  for(size_t i = 0; i < bytes.size(); i++) {
    snprintf(uuid.data() + (i * 2), 2, "%02x", (uint8_t) bytes[i]);
  }

  return uuid;
}
```

Reply:
Also, I removed the hyphenation parameter completely. We don't use it anywhere except tests so why bother keeping it around? If we need it in the future its not like it'd be trivial to add. Also, I've never seen hyphenated versions outside of very legacy uses as those hyphens are dead weight in terms of storage/transmission costs and since random UUIDs aren't parsed there's not reason to pretend there's structure to the value.